스토리 화면 UI 구현

junu0516 opened the pull request:
### 🔖 관련 이슈
- #84 

<br> 

### ⚒️ 작업 내역
- [x] 스토리 화면 UI 구현
- [x] 스토리 화면에 자동 애니메이션 적용
- [ ] Mock Repository로 스토리 화면 기능 구현
<br>

### 💻 리뷰어 가이드
- 홈 화면 상단>아무 친구버튼 터치>터치된 친구의 변기록 노출 의 순서로 동작하면 됩니다.
- 더미데이터 15개의 변 기록이 노출되야 합니다.
- 화면이 나타나면 15개 기록 중 첫번째 기록(인덱스0)에 대해 항상 파란색으로 조회 표시가 나타나야 합니다.
- 화면 우측을 터치하면 다음 기록이 화면에 노출되야 합니다.
- 화면 좌측을 터치하면 이전 기록이 화면에 노출되야 합니다.
- 제일 첫번재 기록(인덱스0)에서 화면 좌측을 터치하면 첫번째 기록이 노출된 상태가 계속 유지되야 합니다.
- 제일 마지막 기록(인덱스 14)에서 화면 우측을 터치하면 마지막 기록이 노출된 상태가 계속 유지되야 합니다.

> 2022.07.04 애니메이션이 적용되서 아래와 같은 애니메이션 동작이 확인되면 됩니다.
- 스토리화면 노출>자동으로 첫번째 기록(인덱스0)에서 시작해서 우측 방향으로 애니메이션이 진행되야 합니다.
- 마지막 기록(인덱스14)까지 자동으로 도달하면 애니메이션이 멈춘 상태로 계속 머물러야 합니다.
- 애니메이션이 멈춘 상태에서 화면 좌측을 터치하면 이전기록이 화면에 노출되고, 애니메이션이 해당 인덱스부터 다시 시작되야 합니다.

<br>

### 📝 부가 설명
- 애니메이션 적용을 위해 처음에 UIView.animate를 사용하다가 좀 삽질을 많이했는데, UIViewPropertyAnimator를 사용하면 애니메이션에 대한 동적인 처리가 더 수월해진단걸 뒤늦게 알아서.. 이걸 사용하는 방식으로 고쳤습니다.
(자세한 내용은 회의 때 공유할께요!)

<br> 

### 📑 첨부 자료

![Simulator Screen Recording - iPhone 14 Pro - 2023-06-28 at 23 07 49](https://github.com/LifePoop/LifePoop_iOS/assets/68586291/5797ee44-37e7-47c2-9a5e-c4e471dc140f)

junu0516 committed:
Merge branch 'develop' into 84-storyScene

junu0516 committed:
#84 chore: 스토리 화면 구현에 필요한 콘텐츠 파일 업데이트

junu0516 committed:
#84 feat: SegmentedProgressView 구현

junu0516 committed:
#84 feat: StoolStoryLogEntity 정의

junu0516 committed:
#84 feat: 스토리 화면 뷰컨트롤러, 뷰모델 구현

junu0516 committed:
#84 feat: 홈화면에서 상단 친구 셀 터치했을 때, 선택된 친구의 변 기록 가져오는 임시로직 구현

junu0516 committed:
#84 feat: 스토리 화면 열고 닫는 로직 구현

junu0516 committed:
Merge branch 'develop' into 84-storyScene

junu0516 committed:
#84 feat: 스토리 화면에 LocalizedString 적용

junu0516 committed:
#84 refactor: 힘주기 완료 문구 흰색으로 나타나도록 수정

junu0516 committed:
#84 feat: Segment의 track, untrack 표시방식 수정

junu0516 committed:
#84 feat: SegmentedProgressView에 자동 애니메이션 기능 추가

junu0516 committed:
#84 feat: 자동 애니메이션 적용에 따라 스토리 화면의 이벤트 방출, 입출력 전달 방식 재수정

junu0516 committed:
Merge branch 'develop' into 84-storyScene

junu0516 committed:
#84 chore: dev 머지 후 컴파일 오류나는 부분 수정

junu0516 committed:
#84 feat: 힘주기 버튼 활성화되는 부분에 애니메이션 적용되지 않도록 처리

junu0516 committed:
#84 feat: 임시로 힘주기 버튼 터치 동작하는 부분 구현

junu0516 committed:
#84 chore: 스토리 시간차 확인을 위해 임시로 StoolLogEntity 더미데이터의 기록 시간 수정

junu0516 committed:
#84 feat: 스토리 화면에 기록된 시간 정보 라벨 추가

임시로 더미데이터 시간 조작해서, 현재 날짜로 업데이트 한 후 현재 시간과의 시간차 구해서 출력하도록 처리함

junu0516 committed:
#84 feat: 스토리화면에서 맨 마지막 기록에 대한 애니메이션 실행 중에 오른쪽 화면 터치하면 화면 닫도록 처리

junu0516 commented:
@sanghyeok-kim 
추가로 기획 수정한거 반영해서 아래 사항 작업 후 푸시했어요 확인 부탁드립니다.
- 맨 마지막 기록에 대한 애니메이션 실행 중에 스크린 우측 부분 터치하면 스토리 화면 닫기도록 처리
- 시간 정보 출력하는 라벨 추가, 임시로 더미데이터 시간 변경

junu0516 committed:
#84 refactor: LifePoopButton의 title 속성을 public var로 변경

junu0516 committed:
#84 feat: 스토리 화면 요소 출력 이벤트 방출 달리하도록 수정, 이전 기록에 대해 버튼 및 라벨 숨김 처리

junu0516 commented:
@sanghyeok-kim  아래 사항 수정해서 추가로 커밋 푸시했어요. 시간 편하실 때 코멘트 부탁드려요
(내용이 좀 있는 것 같아서 회의 때 간단하게 공유드리겠습니다.)

- 스토리 화면에 기록 시간 표시
- 스토리 맨 마지막 기록 애니메이션 실행 중에 스크린 우측 터치하면 스토리 화면 닫기 처리
- 맨 마지막 기록 애니메이션이 끝나면 자동으로 스토리 화면 닫기 처리
- 이전 기록에 대해 응원 문구 및 힘주기 버튼 숨김 처리
- 힘주기 버튼 터치 후 버튼 타이틀 '힘주기 완료'로 머무르도록 처리
- 기타 뷰모델 내부의 이벤트 방출 로직 정리(불필요한 이벤트 방출 안되도록 처리)

junu0516 committed:
#84 refactor: 애니메이션 시간 1초에서 6초로 다시 변경

sanghyeok-kim commented on Projects/Features/FeatureHome/FeatureHomePresentation/Sources/FriendStoolStoryScene/FriendStoolStoryViewModel.swift:
```diff
@@ -0,0 +1,224 @@
+        case right
+    }
+    
+    public struct Input {
+        let viewDidLayoutSubviews = PublishRelay<Void>()
+        let didTapCloseButton = PublishRelay<Void>()
+        let didTapScreen = PublishRelay<ScreenSide>()
+        let didTapCheeringButton = PublishRelay<Void>()
+        let didUpdateProgressState = PublishRelay<Int>()
+    }
+    
+    public struct Output {
+        let stoolStoryLogs = BehaviorRelay<[StoolStoryLogEntity]>(value: [])
+        let shouldUpdateProgressState = PublishRelay<Int>()
+        let shouldUpdateShownStoolLog = PublishRelay<StoolLogEntity>()
+        let shouldHideCheeringButton = PublishRelay<Bool>()
+        let shouldEnableCheeringButton = PublishRelay<Bool>()
+        let shouldUpdateCheeringButtonText = BehaviorRelay<String>(value: LocalizableString.boost)
+        let shouldHideCheeringLabel = PublishRelay<Bool>()
+        let shouldUpdateCheeringLabelText = BehaviorRelay<String>(value: LocalizableString.cheeringWithBoost)
+        let shouldUpdateStoolLogTime = PublishRelay<String>()
+        let shouldUpdateFriendStoolLogSummary = PublishRelay<String>()
+        let shouldShowLoadingIndicator = PublishRelay<Bool>()
+    }
```
Input, Output 프로퍼티의 네이밍 관련해서 수정해주시면 좋을 것 같아요!

1. 저번에 정리한 대로, [명사+동사] 형태로 통일해주세요 (didTapCloseButton -> closeButtonDidTap)
2. 어떤 프로퍼티는 should로 시작하면서 Bool 타입인 반면, 어떤 프로퍼티는 should로 시작하지만 Bool 타입이 아닌 것이 있습니다.
예를 들어 `shouldUpdateCheeringButtonText`의 경우 'CheeringButtonText를 변경해야하는가?'로 해석이 되어 Bool 타입을 가질 것 같지만, 실제로는 String 타입을 가지는 것이 조금 어색하게 느껴집니다. 그냥 `updateCheeringButton`으로 지어도 괜찮을 것 같은데, 어떻게 생각하시나요?

junu0516 replied:
아 네네 1번은 저렇게 통일하면 될 것 같고,
    
2번같은 경우에는 should를 쓴건 '`~를 해야 하는가`'가 아니라 "`~를 해야 한다`"로 의미를 전달하려 한 건데 혼동이 있나 보군요.
input은 대부분 특정 행위 혹은 사건이 이미 발생한 경우기 때문에 did(과거 행위에 대한 조동사)+동사원형(행위) 표현이고, output은 특정 행위가 발생함에 따라, 특정 처리가 반드시 발생해야 하기 때문에 should+동사원형 형태로 통일했어요.    
    
우선은 조동사 자체를 없애는게 이해하기 더 편하신건가요??

sanghyeok-kim replied:
아하 그런 이유가 있었군요.

저는 should를 '~해야하는가?'로 해석되면서 Bool 타입으로 사용할 때에 사용하는게 적절하다고 생각하고,

Output 구조체 안에 들어온 프로퍼티인 이상
이 것을 구독하는 입장에서 'Output에서 시키는대로(방출한대로) 해야 한다'는 의미가 이미 내포되어 있다고 생각해서,
굳이 모든 Output의 프로퍼티에 should를 붙일 필요는 없다고 생각해요

junu0516 replied:
음 넵 그러면 저대로 고쳐서 다시 올릴께요~!

sanghyeok-kim commented on Projects/DesignSystem/Sources/Component/SegmentedProgressView/SegmentedProgressView.swift:
```diff
@@ -0,0 +1,164 @@
+    }
+}
+
+private extension SegmentedProgressView {
+    
+    func trackSegment(forIndexOf targetIndex: Int) {
+        guard segments.count == numberOfSegments else {
+            fatalError("The size of the segments array must be equal to number of segments value in dataSource.")
+        }
+        
+        guard targetIndex < numberOfSegments else {
```
이렇게 UI 관련 로직에서 실패하는 경우에 대비해 DesignSystem 모듈에도 Logger 모듈에 대한 의존성을 추가해 뒀습니다
따라서 fatalError 보다는 Logger를 통해 로그를 출력하고, 함수를 리턴시키는 건 어떨까요?

junu0516 replied:
이건 처음에 로그 찍고 넘어가려다가, 애니메이션이 정상적으로 실행되야 하는 상황에서 단순히 로그만 찍고 넘어가면 화면 흐름 자체도 이상해지고 어떤 에러가 발생한건지 상황을 제대로 인지하기 힘들기 때문에 아얘 런타임 에러를 발생시키는게 낫다고 생각들어요

컬렉션뷰 같은 경우에도 인덱스 관련 에러는 런타임 에러로 바로 처리하는데, 이와 비슷한 맥락으로 생각했어요.

경우에 따라서는 어떤 에러의 경우에는 예외처리 보다는 아얘 런타임 에러를 발생시키는게 맞다고 생각드는데 어떻게 보세요?

sanghyeok-kim replied:
로그만 찍고 리턴시켰다가 사용자 입장에서 애니메이션이 동작하지 않는다면 그거대로 문제고,
인덱스 관련 에러는 컬렉션뷰 처럼 엄격하게 처리되어야하는게 맞으니
코드상 인덱스가 잘못 지정될 가능성이 적다면 말씀하신대로 런타임 에러를 발생시키는 것도 괜찮아 보이네요

sanghyeok-kim commented on Projects/Features/FeatureHome/FeatureHomePresentation/Sources/FriendStoolStoryScene/FriendStoolStoryViewModel.swift:
```diff
@@ -0,0 +1,224 @@
+            .withUnretained(self)
+            .map { `self`, index in
+                let stoolLog = stoolStoryLogs[index].stoolLog
+                let isCheeringUpAvailable = stoolStoryLogs[index].isCheeringUpAvailable
+                let shouldHideCheeringLabel = index < stoolStoryLogs.count-1
+                let shouldHideCheeringButton = index < stoolStoryLogs.count-1
+                let shouldUpdateStoolLogTime = self.getTimeDifference(fromDateOf: stoolLog.date ?? "")
```
Lint Rule에 operator_usage_whitespace가 누락되어서 경고가 발생하지 않았더라구요.
현재 브랜치에 해당 규칙을 추가해 놓을테니, 경고가 발생하는 코드들을 수정해주시면 됩니다!

junu0516 replied:
네 lint 한 번 다시 적용해서 고쳐놓을께요

sanghyeok-kim committed:
#84 docs: Swiftlint에 연산자 사이 공백 규칙 추가

sanghyeok-kim commented on Projects/Features/FeatureHome/FeatureHomePresentation/Sources/FriendStoolStoryScene/FriendStoolStoryViewController.swift:
```diff
@@ -0,0 +1,226 @@
+        button.setTitle(LocalizableString.boost, for: .normal)
+        button.setTitle(LocalizableString.doneBoost, for: .disabled)
+        return button
+    }()
```
loading indicator가 돌아갈 때 title과 indicator가 함께 나타나서 어색한 부분이 있는데요,
여기 코드에서 disabled일 때에 대한 title을 지정해둬서 그런 것 같습니다.

이 부분은 작성하신 코드 그대로 사용할 수 있게 제가 LifePoopButton을 수정해서 고쳐놓을게요
#100 


https://github.com/LifePoop/LifePoop_iOS/assets/57667738/a8ba3754-2086-4c2e-b461-51869ddc25d8

junu0516 replied:
아 네네 말하신대로 수정해주시면 될 것 같아요 감사합니다!

sanghyeok-kim committed:
#100 String+nsAttributedString extension 구현

sanghyeok-kim committed:
#100 refactor: LifePoopButton 코드 리팩토링

- title 설정을 기본 setTitle(for:) 메소드를 사용할 수 있도록 수정
- loading indicator와 disabled 상태에서의 title이 함께 나타나지 않도록 수정

sanghyeok-kim committed:
#100 refactor: 수정된 LifePoopButton 로직에 맞게 코드 수정

sanghyeok-kim committed:
Merge pull request #101 from LifePoop/100-fix-lifepoop-button-loading-indicator

LifePoopButton이 비활성화 상태일 때 loading indicator와 title이 함께 나타나는 문제 해결

junu0516 committed:
Merge branch 'develop' into 84-storyScene

junu0516 committed:
#84 chore: 뷰모델 input, output 변수명 수정, lint 경고 제거

junu0516 committed:
#84 feat: 홈 화면 상단 친구 목록에서 인덱스 0에 대한 셀 터치 제외 임시로 추가
